torch/lib/THCUNN/SoftMaxCommon.cuh

```diff
@@ -0,0 +1,454 @@
+        for (uint32_t d = 0; d < dim_size; d++)
+          output[data_offset + d * dim_stride] = epilogue(input[data_offset + d * dim_stride]);
+      }
+    }
+  }
+}
+
+template <typename T, typename AccumT, template<typename, typename> class Epilogue>
+__global__ void cunn_SpatialSoftMaxBackward(
+    T *gradInput, T *output, T *gradOutput,
+    uint32_t outer_size, uint32_t dim_size, uint32_t inner_size)
+{
+  SharedMem<AccumT> smem;
+  const uint32_t outer_stride = inner_size * dim_size;
+  const uint32_t dim_stride = inner_size;
+
+  for (uint32_t outer_index = blockIdx.x; outer_index < outer_size; outer_index += gridDim.x) {
+    const uint32_t outer_offset = outer_index * outer_stride;
+    for (uint32_t inner_index = blockIdx.y * blockDim.y + threadIdx.y; inner_index < inner_size; inner_index += blockDim.y * gridDim.y) {
+      const uint32_t data_offset = outer_offset + inner_index;
+      // See the comment in forward kernel
+      if (blockDim.x > 1) {
+        AccumT sum = 0;
+        for (uint32_t d = threadIdx.x; d < dim_size; d += blockDim.x)
+          sum += gradOutput[data_offset + d * dim_stride];
+        sum = spatialBlockReduceX<AccumT, Add>(smem.getPointer(), sum);
+
+        Epilogue<T, AccumT> epilogue(sum);
+        for (uint32_t d = threadIdx.x; d < dim_size; d += blockDim.x) {
+          gradInput[data_offset + d * dim_stride] =
+            epilogue(gradOutput[data_offset + d * dim_stride],
+                    output[data_offset + d * dim_stride]);
+        }
+      } else {
+        AccumT sum = 0;
+        for (uint32_t d = 0; d < dim_size; d++)
+          sum += gradOutput[data_offset + d * dim_stride];
+
+        Epilogue<T, AccumT> epilogue(sum);
+        for (uint32_t d = 0; d < dim_size; d++) {
+          gradInput[data_offset + d * dim_stride] =
+            epilogue(gradOutput[data_offset + d * dim_stride],
+                    output[data_offset + d * dim_stride]);
+        }
+      }
+    }
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Regular kernel (fast when dim_size is large; requires inner_size == 1)
+////////////////////////////////////////////////////////////////////////////////
+
+
+template <typename T, typename AccumT>
+struct MaxFloat
+{
+  __device__ __forceinline__ AccumT operator()(AccumT max, T v) const {
+    return fmaxType(max, v);
+  }
+};
+
+template<typename T, typename AccumT>
+struct AddFloat
+{
+  __device__ __forceinline__ AccumT operator()(AccumT sum, T v) const {
+    return sum + v;
+  }
+};
+
+template<typename T, typename AccumT>
+struct SumExpFloat
+{
+  __device__ __forceinline__ SumExpFloat(T v)
+    : max_k(v) {}
+
+  __device__ __forceinline__ AccumT operator()(AccumT sum, T v) const {
+    return sum + THCNumerics<T>::exp(v - max_k);
+  }
+
+  const T max_k;
+};
+
+template <template<typename> class Reduction, typename AccumT>
+__device__ __forceinline__ AccumT
+blockReduce(AccumT* smem, AccumT val,
```

Comment:
Pour one out for https://github.com/torch/cutorch/pull/750

Reply:
That's why I asked about it